src/utils/EventEmitterBasedState.ts

```diff
@@ -0,0 +1,46 @@
+import { Mutex } from "await-semaphore";
+import { isEqual } from "underscore";
+import { Disposable, EventEmitter } from "vscode";
+
+export default class EventEmitterBasedState<T> {
```

Comment:
i think it should be disposable bc EventEmitter is disposable.

moreover EventEmitterBasedState can extend  EventEmitter  instead of composition